### ISSUE TITLE
Error Creating Shipment

### ISSUE THREAD

**Issue by hverdonk** [vsoch/freegenes#52]
**Describe the bug**
When creating shipment for [this order](https://freegenes.dev//o/details/cd3f748d-058f-43f6-8e23-3fa1c3537f7c), the shipment did not appear to process, and I received an error message.

**To Reproduce**
From the linked order, click "create shipment." Fill out all the fields in the form, selecting "contains dry ice," then click submit.

**Expected behavior**
The webpage should let me select a shipping service based on the addresses I've added, and let me add my PTA number in the equivalent to Shippo's Reference Field 1.

If the site defaults to using FedEx as the shipping service, the problem may be that FedEx requires an email and phone number for both the shipper and the receiver.

If applicable, add versions and screenshots to help explain your problem.

The error page produced this as well:

<redacted>

**Comment by vsoch** [comment]
It actually looks like the submission was successful (see status QUEUED and that it's not a test token) but there was likely another server error. I'm still in Colorado but I'll take a look end of week!

**Comment by vsoch** [comment]
Also - it's not you adding the PTA number - this is already implemented on the server. Your user interaction should be the same as it was before.

**Comment by hverdonk** [comment]
Gotcha, thank you. I have a few comments about the user interaction, but I'll let you get back to your vacation!

**Comment by vsoch** [comment]
hey @hverdonk I realized this bug could slow down lab operations, so I wanted to take a look asap!

It looks like everything actually worked, and the bug was that we were looking for a status of "SUCCESS" to determine that the shipment was created. Actually, with the live key it returned a QUEUED status. What I just did is to update the view to allow success or queued - if you want to give it another try. Also note that I've redacted the json you originally provided, just so we don't share addresses, and that you still won't need to enter the reference stuffs - this is from the (now removed) json to show that we have the reference_1 defined.

```
   "extra":{
      "dry_ice":{
         "contains_dry_ice":true,
         "weight":"2"
      },
      "reference_1":"11111111-1-AAAAA"
   },
```
If a phone is required (and we don't ask for it) that might trigger a bug when you select the rate - let's try out this step and see if we trigger that as an issue!

**Comment by hverdonk** [comment]
Thanks for redacting the json - I didn't notice it contained an address. Speaking of addresses, the "Create Shipment" form doesn't have a section for City/State. I've been using the "Shipping address2" field for now, but I'm not sure it will produce the expected behavior - Shippo has both a "Shipping Address 2" field for addresses with apartment numbers and two separate fields for city and state/province.

Right now, the site will proceed to the rate selection page, but is only offering USPS shipping rates, so I can't check if FedEx errors on it or not.

**Comment by vsoch** [comment]
A few quick notes! 

The state and city would be redundant - because we can derive that from the zip code. For example, here is an address that you input - notice how State and City are defined, and you added them as another line for Street 2 (not necessary)

```
      "city":"Stanford",    # CITY
      "company":"Stanford University",
      "country":"US",
      "email":"freegenes@stanford.edu",
      "is_complete":true,
      "is_residential":false,
      "name":"Marc Fleischmann",
      "object_id":"487b0086bc9e4e8faa8eca5455dfdc9c",
      "phone":"",
      "state":"CA",   # STATE
      "street1":"443 Via Ortega Rm 252",
      "street2":"Stanford Ca",  # NOT NECESSARY
      "street3":"",
      "street_no":"",
      "test":false,
      "validation_results":{
```

I'm not privy to how the options are selected based on your criteria, but (at least from the first json you posted) the options were also USPS only. If we take a look at the json again, there is also a warning message about DHL express:

```
   "messages":[
      {
         "code":"",
         "source":"DHLExpress",
         "text":"Shippo's DHL Express master account doesn't support shipments to inside of the US"
      }
   ],

```
You mentioned that phone number is required for FedEx, for both receiver and shipper? I think we should try adding that (can you send me via email a lab phone number, and I'll have it as a variable on the server, and then update the shipping form to ask for a phone number for the recipient?)

After we get phone numbers in we can test again. If it's still the case that FedEx doesn't show up, then we should step back and try removing the extra reference, and then (if still no go) the extra ice metadata. 

It's not that FedEx is erroring (there aren't any errors) it's moreso that Shippo is filtering down options based on inputs, and we are providing (or not providing) some inputs that make it not an option.

**Comment by vsoch** [comment]
okay! So I grabbed Marc's phone, added to the server, and added a field for the checkout form to add a phone number:

![image](https://user-images.githubusercontent.com/814322/65272688-87164a00-daed-11e9-860a-58079a60a8e7.png)

If you have this particular lab phone (it should have been provided as it's a required field on the checkout form) then you should now be able to put it into the shipping form. Give that a try as our next debugging step. If you don't see options for FedEx, try removing dry ice. If that doesn't work, then I'd like to try making the request on the server (so I can inspect output) - send me an update when you've given the changes a try!

**Comment by hverdonk** [comment]
I added a phone number, and FedEx appeared! I bought a label for [one of the dummy shipments](https://freegenes.dev/o/details/cd3f748d-058f-43f6-8e23-3fa1c3537f7c) being sent to nobody, and it gave me a FedEx error that the package dry ice weight couldn't be more than the actual package weight. The error was totally my fault - I accidentally put the wrong numbers in when filling out the weights - but now I can't go back and fix them, and the order is listed as "complete" in the system.

Also, when I reached the label purchasing screen for [this shipment](https://freegenes.dev/o/details/afd3dd30-3436-4b3e-a265-562174f3bfac), I clicked away without purchasing anything since it didn't yet have FedEx as an option, intending to return to it later. However, it now has a label associated with it, and is also listed as a complete shipment. Digging deeper, I saw that, though the order was created in September, the label corresponds to a much earlier shipment that was completed in July, and the "track package" link reflects that.

Other than that, the system is working pretty well! Thanks for clarifying the City/State thing to me!

**Comment by vsoch** [comment]
Great! I'll follow up on the new issues tomorrow - it sounds like you hit one of the orders that we tried to import (but didn't have complete into for) - I'll look more into it and send you what I learn.

**Comment by hverdonk** [comment]
Sounds great, thanks!

**Comment by vsoch** [comment]
@hverdonk for the first issue, I think the first thing to try would be to prevent you from creating the shipment given that the dry ice weighs more than the parcel. I've added a check (just testing locally) and it will look like this:

![image](https://user-images.githubusercontent.com/814322/65333136-ec217c80-db8d-11e9-9fbb-2bb8bb0b12e1.png)

Note that the default weight for dry ice, per Keoni's scripts, is set to 2 lbs on the server (regardless of the package size). I'm not sure if you need this exposed, but if you hit a case where you need to specify more let me know and we can do that.

For the order - could you describe what you mean by "click away?" The label would be generated by the "Create label" button on the transaction page (which I require you to click for this reason, and since there is a slight delay with the API) so if you didn't actually click that, it's likely that it was an old order that had the label updated. But it looks like it was generated early September, so I'm not sure!

Anyway - I just removed the label for that order, so you should be able to try again. I also better exposed messages from the server to the interface across views, so hopefully debugging will be a bit easier. Give it another try and send me an update!

**Comment by hverdonk** [comment]
I like that the software would prevent me from creating the shipment - it's simple and elegant. I'd prefer the dry ice weight field be exposed, but whether we expose it or not, we're using more dry ice than that at the moment (around 5 lbs).

For the other order, I went to choose a label, saw that the only available options (at the time) were USPS, and closed the tab without clicking "Create Label." The order itself was generated early September, but the package label that it was formerly associated with was marked as early July. Either way, thank you for removing the label! I tried filling the order again, but it's back to only showing me USPS. I **do** have the recipient's phone number listed this time, so I'm not really sure why FedEx isn't being offered.

**Comment by vsoch** [comment]
You triggered a server error - I just fixed and restarted to fix it. I'll add a field for dry ice.

**Comment by hverdonk** [comment]
The server error was probably because I didn't select any dry ice option to see what would happen.

**Comment by vsoch** [comment]
Actually it was something different - a view was missing a return statement so it didn't return anything at all :) 

For the FedEx being available or not - I think it comes down to the kind of addresses you have (reading here -> https://support.goshippo.com/hc/en-us/articles/204610169-Troubleshooting-FedEx-Ground-rates) it seems like residential vs. commercial is important) and also that Fedex is [connected to Shippo](https://support.goshippo.com/hc/en-us/articles/360031117552-Connect-your-FedEx-Account-to-Shippo). If you aren't able to reliably produce the options, I would send in a request to Shippo to ask specifically what criteria trigger showing those options or not. The most I can do is to optimize the request based on this feedback - I have basically no control about the rates that are returned back from their API.

Do you want to send them a support request and update me when you hear?

**Comment by hverdonk** [comment]
Sending them a support request now - I'll update you when they respond. It's worth noting that I doubt we'll be using residential addresses often. Thanks for all your help! :)

**Comment by vsoch** [comment]
hey an error was just triggered, and it's because the form submit was invalid (but I forget to handle this case in the view). I'm working on some factory backend stuffs now but I'll get a fix for the form in the next update.

**Comment by hverdonk** [comment]
Sounds good! I haven't touched the code today, but Keoni thinks that Isaac (on Discord) might have been the one to discover the form submit bug. If you need any more information about what happened, I'd start with him :)

**Comment by vsoch** [comment]
I get full error callbacks via Sentry, so I know exactly what the issue was! Thanks though, and maybe I'll ping him in Discord so he knows.

**Comment by hverdonk** [comment]
Just to keep you updated, I still haven't heard back from Shippo about their criteria for showing/excluding FedEx rates. I emailed them again today though, so hopefully that'll get some attention.

**Comment by vsoch** [comment]
ok thanks for the update! If you don't hear, then I'll go in for a try. Something something... squeaky wheels getting grease! :)

**Comment by hverdonk** [comment]
Here's the response I got from Shippo today:

> Thanks for reaching out, I apologize you're running into trouble with Shippo.
>
>First, I'd like to apologize for the delay in responding. We are working through a bit of a backlog and trying to get back to our customers as quickly as possible. We are working on hiring additional support reps to ensure we are able to meet our customers' needs. Thank you for your patience and understanding.
>
>Regarding the FedEx rating issue, there were other reports of this over the last 4 days, however, I do believe the issue has been resolved. Are you still running into trouble with this?
>
>If so, would you mind sending me the request/response you are receiving when calling the api for rating?

I tried to create the shipment again, but it's still only listing USPS. If you could send me the request/response data, I'd be happy to pass it along.

**Comment by vsoch** [comment]
Yes I can do that! Can you send me (in email) the fields you are providing to the form? I'll log into the server and manually create the requests (painful, but essential!) haha.

**Comment by hverdonk** [comment]
Sent! Thank you!

**Comment by hverdonk** [comment]
I finally received this response from Shippo today:
>Looking at the ship to and ship from addresses I don't see the city and state fields populated. Are >those being added in the request? If not, can you please add those in to see if that resolves the >issue?
If you change up the software to populate those two fields, I can create another dummy shipment and test out whether or not it returns FedEx rates.

**Comment by vsoch** [comment]
@hverdonk I just added city and state, for both the lab and shipping to addresses. The user should be able to input it on the form that goes to the lab email, and then it should be passed to shippo.

**Comment by hverdonk** [comment]
Okay, I did a submission with the address and still no dice. I've reached back out to Shippo and let them know.

**Comment by vsoch** [comment]
Just sent you complete request information via email, I expect Shippo will ask for it.

**Comment by hverdonk** [comment]
I just sent it straight to them, since I'm not waiting another week for them to ask for it.

**Comment by vsoch** [comment]
That's how to do it! :+1:

**Comment by hverdonk** [comment]
Here's Shippo's response:

>Thanks for your patience with this.

When we tested pulling rates with the request you sent over we were able to get them successfully for FedEx.

Our engineer mentioned that your call is passing the /shipment object in their /shipment request. He suggested that you structure your call using this API doc page: https://goshippo.com/docs/shipping-labels/

You can also reference this API reference doc for adding the add'l items like dry ice: https://goshippo.com/docs/reference#shipments-extras

Once you format the calls in this way you shouldn't run into any further issues, however, if you do it will be much easier to spot where they are coming from.

**Comment by vsoch** [comment]
Hmm, so there are a couple issues with their response - firstly, the object that I sent them is the complete response for the request - the actual request is indeed created (exactly) according to their example documentation. If the engineer wants to take a look to confirm this, it's [all right here](https://github.com/vsoch/freegenes/blob/8073fc35e8353ad60883349535ad064f6d99aede/fg/apps/orders/views/shipping.py#L230).

In that they were able to make the request and get the FedEx rates, that suggests there might be issue with our token (all else the same). Could we ask for them to test making the request with our token? Or checking permissions / scope for it?

To summarize - our calls are already formatted as their have specified, and so the difference in behavior between their test and our s isn't logical. The only variable that comes into question is thus the shipping token.

**Comment by hverdonk** [comment]
Here is what I responded to them with. It's nearly word-for-word with your post:

> Thanks for updating me. The code object that I sent you is the complete response for the request - the actual request is indeed created (exactly) according to your example documentation. If the engineer wants to take a look to confirm this, it's all right here.
> 
> Given that you were able to make the request and get the FedEx rates, that suggests there might be an issue with our token (all else the same). Would you please test making the request with our token? Or checking permissions / scope for it?

Personally, my theory is that they took the address, put it in their desktop site, saw FedEx rates, and assumed that validation was good enough. Hopefully, forcing them to test the API token will force them to look at the code itself if that's the case.

**Comment by vsoch** [comment]
Yes, and if they respond in the same way and it’s not helpful, what I’ll do is create an isolated script that generate an address / shipment exactly as we do, and then create an asciinema recording of me running it on the server and showing them the actual output. Their “but it works for me” response really doesn’t work for us, so we’ll just work a little harder to give them exactly something to reproduce, and a recording of us running it.

**Comment by hverdonk** [comment]
With repeated poking, I got this back from Shippo:

> Thank you for reaching back out. I apologize for the delay here.
> 
> Looking at your most recent request, it appears you are not adding a phone number and this is an international shipment. Phone numbers are required for international labels and you will return an error if it's not listed. Would you mind trying to add a phone number to your request and see if that helps? 
> 
> Also, the address_to field may be too long and be causing the error. I see that in your latest request the address_to is listed as "The Chancellor, Masters, and Scholars of the University of Cambridge". Do you ming shortening this to see if that helps?

**Comment by vsoch** [comment]
I think they are punting here- because the request example that we sent them to Isaac Larkin

 - did have a phone number
 - is not international
 - did not have an address_to that was long.

**Comment by vsoch** [comment]
This is really getting us nowhere, I'm going to prepare a single top to bottom script to perfectly reproduce the issue, and record an asciinema of me running it so they can't claim we are getting something different.

**Comment by hverdonk** [comment]
I'd agree with that. I only got this response today because I emailed Shippo yet again to check in and let them know I wanted results (in more polite language). I'll send your content to Shippo and escalate Shippo's poor customer service issue with my coworkers.

**Comment by vsoch** [comment]
okay, I'll update you here with files / links once I generate them - I think we shouldn't escalate until after that.

**Comment by hverdonk** [comment]
Got it, I'll hold off on escalation.

Taking another look at Shippo's response, I'm not sure where they're getting the address that they're testing. "The Chancellor, Masters, and Scholars of the University of Cambridge" isn't one of the recent shipments I've made using the current Shippo token. Can you double check that the software is using the correct token everywhere in the code?

**Comment by vsoch** [comment]
We only have two tokens, one for test (which wouldn't create a shipment) and one live, both which were sent by Keoni. If Shippo is telling you about a wrong shipment, they might be looking at the wrong account. I would check with them and @Koeng101 to make sure that the account they think we have is the one that we do have! And of course don't post tokens anywhere that isn't encrypted / secure.

**Comment by vsoch** [comment]
okay sent you an email with private gist and video - beyond providing them with a script to exactly reproduce what we are getting, I'm not sure what more we can do.

**Comment by hverdonk** [comment]
I've passed it along. I'll keep you updated about Shippo.

**Comment by vsoch** [comment]
The saga continues! :space_invader:

**Comment by hverdonk** [comment]
Alright! Now that we're seeing FedEx rates, I propose that we close this particular issue and I'll make individual reports about any other bugs/problems that arise.

**Comment by vsoch** [comment]
Yes 100% And I'll link the issue I opened with the Shippo API client for future people trying to debug similar. https://github.com/goshippo/shippo-python-client/issues/56
